### PR TITLE
Represent the invalid dates in DatePicker in grey color.

### PR DESCRIPTION
**Pull request by rajat4064g**
#### WHAT
Make the invalid dates (dates before fromDate and after toDate) visible in DatePicker but represent them in grey color.
This also solves the retaining of previous day value in the day picker if month picker is scrolled.

#### Checklist :clipboard:
- [ ] Add explicit visibility modifier and explicit return types for public declarations
- [x] Run spotless check
- [x] Run tests
- [x] Update metalava's signature text files

### CODE REVIEW

**Comment by yschimke** [comment]
Can you add a test to com.google.android.horologist.screensizes.DatePickerTest that covers this?

Would be good to see the result visually.

But I'm probably not the right person to review.

File: composables/src/main/java/com/google/android/horologist/composables/TimePicker.kt [hunk 1]
```diff
@@ -588,19 +589,23 @@ internal fun pickerGroupItemWithRSB(
     )
 }
 
-internal fun pickerTextOption(textStyle: TextStyle, indexToText: (Int) -> String):
-    (@Composable PickerScope.(optionIndex: Int, pickerSelected: Boolean) -> Unit) = { value: Int, pickerSelected: Boolean ->
+internal fun pickerTextOption(
+    isValid: (Int) -> Boolean = { true },
```
Comment (BowerSteve) [inline composables/src/main/java/com/google/android/horologist/composables/TimePicker.kt:593]: Maybe add isValid as the last parameter (it's optional), then textStyle wouldn't need to be specified as a named parameter?

File: composables/src/main/java/com/google/android/horologist/composables/DatePicker.kt [hunk 2]
```diff
@@ -332,6 +334,11 @@ public fun DatePicker(
                         }
                         .focusRequester(focusRequesterConfirmButton)
                         .focusable(),
+                    enabled = if (pickerGroupState.selectedIndex >= 2) {
+                        datePickerState.isDayValid()
```
Comment (BowerSteve) [inline composables/src/main/java/com/google/android/horologist/composables/DatePicker.kt:338]: Seems wrong - shouldn't it be checking for valid month and year as well?

File: composables/src/main/java/com/google/android/horologist/composables/DatePicker.kt [hunk 3]
```diff
@@ -413,72 +415,105 @@ internal class DatePickerState constructor(
...
     private val toDate: LocalDate?,
 ) {
-    private val yearOffset = fromDate?.year ?: 1
-    val numOfYears = (toDate?.year ?: 3000) - (yearOffset - 1)
     val yearState =
         PickerState(
-            initialNumberOfOptions = numOfYears,
-            initiallySelectedOption = date.year - yearOffset,
+            initialNumberOfOptions = 200,
+            initiallySelectedOption = date.year - 1901,
```
Comment (BowerSteve) [inline composables/src/main/java/com/google/android/horologist/composables/DatePicker.kt:421]: Worth commenting that the year range 1900-2100 has come from b/277885199 (and should it be 201 years to include both ends of the range?)

**Review by BowerSteve — COMMENTED** [review]
Please add from-date and to-date samples into the Horologist sample app.

File: composables/src/main/java/com/google/android/horologist/composables/DatePicker.kt (hunk 3 above)
Comment (rajat4064g) [reply composables/src/main/java/com/google/android/horologist/composables/DatePicker.kt]: Done

File: composables/src/main/java/com/google/android/horologist/composables/DatePicker.kt (hunk 2 above)
Comment (rajat4064g) [reply composables/src/main/java/com/google/android/horologist/composables/DatePicker.kt:341]: `datePickerState.isDayValid()` internally checks validity of month and year as well.

File: composables/src/main/java/com/google/android/horologist/composables/TimePicker.kt (hunk 1 above)
Comment (rajat4064g) [reply composables/src/main/java/com/google/android/horologist/composables/TimePicker.kt]: Done

**Pull request merged**